### PR TITLE
Fixes and new benchmark.

### PR DESCRIPTION
**Pull request by equationalapplications**
## WikiMemory read-path fix

Narrows the dimension-mismatch fallback in `WikiMemory.read()` to ignore zero-weight (non-scored) entities, so multi-entity reads no longer fall back to keyword search due to stale embeddings in entities that carry no weight. Also tightens `tierWeights` sanitization exposure for single-entity reads.

- `packages/core/src/WikiMemory.ts` — dimension-mismatch check scoped to scored entities only
- `packages/core/__tests__/multiEntityVectorRanker.test.ts` — zero-weight entity fallback test
- `packages/core/__tests__/multiEntityRead.test.ts` — single-entity `tierWeights` edge case test

## FinanceBench retrieval benchmark

Adds a financial-domain retrieval benchmark alongside the existing SciFact BEIR benchmark. Measures Hit Rate@5, Hit Rate@10, and MRR@10 across the 150-question open-source FinanceBench sample (`PatronusAI/financebench`). Asserts MRR@10 ≥ 0.55 as a regression gate.

**Spec:** [`docs/superpowers/specs/2026-05-13-financebench-integration.md`](https://github.com/equationalapplications/expo-llm-wiki/blob/kv/updates/docs/superpowers/specs/2026-05-13-financebench-integration.md)

New files (all in `packages/integration`):
- `helpers/mrr.ts` — `computeMRR` and `computeHitRate` helpers
- `scripts/fetch-financebench.ts` — fetches corpus from Hugging Face datasets API
- `scripts/embed-financebench.ts` — embeds with BGE-small-en-v1.5, emits frozen fixtures
- `__tests__/financebench.test.ts` — benchmark test with MRR@10 ≥ 0.55 assertion
- `fixtures/financebench-{dump,embeddings,queries,qrels}.*` — frozen fixture files

**Results (first run, 2026-05-13, maxResults=32):**

| Metric | Value | Threshold |
|--------|-------|-----------|
| MRR@10 | **0.7017** | ≥ 0.55 |
| Hit Rate@5 | **0.9067** | — |
| Hit Rate@10 | **0.9667** | — |

### CODE REVIEW

**Comment by equationalapplications** [comment]
## /fix-pr follow-up

**Commit:** `a1e15a9`

### Review resolution

- **Thread: empty-qrel silent skip / inflated metrics** (`financebench.test.ts:92`) — **Fixed:** Added pre-loop assertion `expect(noRelevant).toHaveLength(0)` that fails loudly if any query has no relevant docs; added post-loop `expect(mrrScores).toHaveLength(Object.keys(queries).length)` to confirm all 150 queries were scored. The `continue` guard remains as dead-code defence.
- **Thread: spec says "No changes to packages/core"** (`specs/2026-05-13-financebench-integration.md:96`) — **Fixed:** Replaced that line with a note accurately describing the core changes (`WikiMemory.ts` dimension-mismatch fix + two new core test files) and their relationship to this PR.
- **Thread: corpus fixture size conflict (~90 KB vs ~400 KB)** (`benchmark-results doc:47`, `spec:549`) — **Fixed:** Measured actual file size (269 KB); updated benchmark doc from ~90 KB → ~270 KB and spec from ~400 KB → ~270 KB.

### Verification

- `npm run typecheck` — **pass** (all 4 packages)
- `npm run lint` — **N/A** (no lint script in root or packages)
- `npm run test` (core unit tests) — **pass** (302 tests, 30 files). Full integration benchmark (`financebench.test.ts`) has a 5-min ONNX setup; not re-run in this fix pass — it was green on the prior commit.

File: packages/integration/__tests__/financebench.test.ts [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    'MRR@10 ≥ 0.30 across all 150 FinanceBench questions',
+    async () => {
+      const mrrScores: number[] = [];
+      const hitAt5: number[] = [];
+      const hitAt10: number[] = [];
+
+      for (const [queryId, queryText] of Object.entries(queries)) {
+        const relevant = new Set(qrels[queryId] ?? []);
+        if (relevant.size === 0) continue;
+
```
[the remark this answers is not in the document]
Comment (equationalapplications) [reply packages/integration/__tests__/financebench.test.ts:99]: Fixed in a1e15a9: added `expect(noRelevant).toHaveLength(0)` before the loop (fails loudly if any fixture query has no qrels) and `expect(mrrScores).toHaveLength(Object.keys(queries).length)` after the loop to verify all 150 queries were scored.

File: docs/superpowers/specs/2026-05-13-financebench-integration.md [hunk 2]
```diff
@@ -0,0 +1,585 @@
...
+│   ├── financebench-queries.json    ← generated by fetch script
+│   ├── financebench-qrels.json      ← generated by fetch script
+│   ├── financebench-dump.json.gz    ← generated by embed script
+│   └── financebench-embeddings.json.gz ← generated by embed script
+└── __tests__/
+    ├── scifact.test.ts              (existing)
+    └── financebench.test.ts         ← NEW
+```
+
+No changes to `packages/core` or any other package.
```
[the remark this answers is not in the document]
Comment (equationalapplications) [reply docs/superpowers/specs/2026-05-13-financebench-integration.md]: Fixed in a1e15a9: replaced the inaccurate statement with a note describing the actual core changes — `WikiMemory.ts` dimension-mismatch fix scoped to zero-weight entities, plus two new core test files (`multiEntityVectorRanker.test.ts`, `multiEntityRead.test.ts`).

File: docs/benchmarks/2026-05-13-financebench-retrieval-benchmark.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+### Fixtures
+
+| File | Size | Contents |
+|------|------|----------|
+| `fixtures/financebench-corpus.jsonl` | ~90 KB | Raw corpus (gitignored, re-generate with fetch script) |
+| `fixtures/financebench-queries.json` | ~30 KB | 150 question strings keyed by `financebench_id` |
+| `fixtures/financebench-qrels.json` | ~15 KB | Relevant fact IDs per query |
+| `fixtures/financebench-dump.json.gz` | 0.1 MB | WikiMemory text-only dump (no blobs) |
+| `fixtures/financebench-embeddings.json.gz` | 0.6 MB | Precomputed 384-dim float32 vectors |
```
[the remark this answers is not in the document]
Comment (equationalapplications) [reply docs/benchmarks/2026-05-13-financebench-retrieval-benchmark.md:47]: Fixed in a1e15a9: measured actual file size (269 KB). Updated benchmark doc from ~90 KB → ~270 KB and spec from ~400 KB → ~270 KB so both docs agree.

**Comment by equationalapplications** [comment]
## /fix-pr follow-up

**Commit:** `bdb1db3`

### Review resolution
- **Thread: empty qrels skip / query count** (`financebench.test.ts:97`) — **Fixed:** added `expect(Object.keys(queries).length).toBe(150)` and `expect(Object.keys(qrels).length).toBe(150)` before the benchmark loop; truncated fixtures now fail loudly.
- **Thread: spec "No changes to packages/core"** (`specs/2026-05-13-financebench-integration.md:1`) — **Already resolved** in `a1e15a9` (prior commit).
- **Thread: fixture size estimate conflict** (`docs/benchmarks/…-benchmark.md:47`) — **Already resolved** in `a1e15a9` (prior commit).
- **Thread: mrrScores length consistency** (`financebench.test.ts:93`) — **Already resolved** in `a1e15a9` (prior commit).

### Verification
- `npm run typecheck` — pass
- `npm run lint` — N/A (no lint script at workspace root)
- `npm run test` (core unit tests, 302 tests) — pass

File: packages/integration/__tests__/financebench.test.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        const { facts } = await wiki.read(ENTITY_ID, queryText);
+        const rankedIds = facts.map((f) => f.id);
+
+        mrrScores.push(computeMRR(rankedIds, relevant, 10));
+        hitAt5.push(computeHitRate(rankedIds, relevant, 5));
+        hitAt10.push(computeHitRate(rankedIds, relevant, 10));
+      }
+
+      expect(mrrScores).toHaveLength(Object.keys(queries).length);
+
```
[the remark this answers is not in the document]
Comment (equationalapplications) [reply packages/integration/__tests__/financebench.test.ts:109]: Fixed in bdb1db3: added `expect(Object.keys(queries).length).toBe(150)` and `expect(Object.keys(qrels).length).toBe(150)` before the benchmark loop so truncated fixtures fail loudly rather than silently passing.

**Comment by equationalapplications** [comment]
## /fix-pr follow-up

**Commit:** `b935700`

### Review resolution
- **Thread: "three configurations" vs four rows (tuning doc L3)** — **Fixed:** Changed "three" → "four" in `docs/benchmarks/2026-05-13-financebench-retrieval-tuning.md`
- **Thread: doc floor/config mismatch (tuning doc L44, L80 vs test)** — **Fixed:** `packages/integration/__tests__/financebench.test.ts` — both `WikiMemory` instances updated to `maxResults: 32`; assertion raised from `>= 0.30` to `>= 0.55`; test name updated to "MRR@10 ≥ 0.55". Code now matches the tuning doc's stated final configuration.

### Verification
- `npm run typecheck` — pass (all 4 packages)
- `npm run lint` — N/A (no lint script in repo)
- `npm run test` — pass (core: 302 tests, integration non-benchmark: 37 tests; financebench benchmark skipped to avoid ONNX model download in local run)

File: docs/benchmarks/2026-05-13-financebench-retrieval-tuning.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
+# FinanceBench retrieval tuning — 2026-05-13
+
+Follow-up to the [initial FinanceBench benchmark](./2026-05-13-financebench-retrieval-benchmark.md). Evaluated three configurations on the same frozen fixture set to determine the effect of `hybridWeight`, `preFilterLimit`, and `maxResults`.
```
[the remark this answers is not in the document]
Comment (equationalapplications) [reply docs/benchmarks/2026-05-13-financebench-retrieval-tuning.md]: Fixed in b935700: changed "three" → "four" in the intro paragraph — the table correctly listed four configurations (Baseline + three new variants), but the text undercounted.

File: docs/benchmarks/2026-05-13-financebench-retrieval-tuning.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+|--------|--------|-----------|------------|------------|
+| Baseline (semantic, maxResults=10) | 0.7017 | 0.9067 | 0.9667 | — |
+| **Guided: hw=0.5, preFilterLimit=1000, maxResults=32** | **0.2679** | **0.4133** | **0.6067** | **0.9000** |
+| **Semantic-32 (winner)** | **0.7017** | **0.9067** | **0.9667** | **0.9933** |
+| hw=0.9, maxResults=32 | 0.6283 | 0.8600 | 0.9400 | 1.0000 |
+
+## Final asserted floor
+
+`MRR@10 ≥ 0.55` (semantic-32 config, actual 0.7017 — 1.27× floor).
+
```
[the remark this answers is not in the document]
Comment (equationalapplications) [reply docs/benchmarks/2026-05-13-financebench-retrieval-tuning.md:44]: Fixed in b935700: both WikiMemory instances in the test now use maxResults: 32; assertion raised from >= 0.30 to >= 0.55; test name updated to match. The test now reflects the semantic-32 final config described in this doc.

File: docs/superpowers/specs/2026-05-13-financebench-integration.md [hunk 7]
```diff
@@ -0,0 +1,620 @@
...
+The SciFact BEIR benchmark validates retrieval quality on scientific abstracts. There is no corresponding benchmark for **dense financial document retrieval** — the domain most relevant to apps like Curated Thoughts that ingest SEC filings, 10-Ks, earnings releases, and similar regulatory documents.
+
+Without a financial domain benchmark, regression in retrieval quality on financial language goes undetected. BGE-small-en-v1.5 might score acceptably on SciFact while silently degrading on the dense, table-heavy, numeric-rich text found in financial filings.
+
+---
+
+## Requirement
+
+**Add a FinanceBench retrieval benchmark to `packages/integration` that measures Hit Rate@5, Hit Rate@10, and MRR@10 across the 150-question open-source FinanceBench sample. Assert MRR@10 ≥ 0.30. Emit a benchmark-results JSON report. Follow the same fixture/script/test structure as the existing SciFact benchmark.**
+
```
[the remark this answers is not in the document]
Comment (equationalapplications) [reply docs/superpowers/specs/2026-05-13-financebench-integration.md:19]: Fixed in 8b42e84: updated spec Requirement, Goals, code snippet (maxResults 10→32, ≥0.30→≥0.55), and Pass/Fail Threshold table to match the current CI gate (MRR@10 ≥ 0.55, maxResults=32). Retained a note that 0.55 was raised from the original 0.30 floor.

File: docs/benchmarks/2026-05-13-financebench-retrieval-benchmark.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+| **Vector width** | 384 |
+
+Corpus embeddings are **frozen** in `fixtures/financebench-embeddings.json.gz` and restored directly into `embedding_blob` before the test runs. Only the 150 query embeddings are computed at test time via `FlagEmbedding.init`.
+
+## What was tested
+
+| Suite | Corpus | Queries | Metrics | Asserted floor |
+|-------|--------|---------|---------|----------------|
+| **FinanceBench** | 180 unique evidence texts from 150 10-K/earnings filings | 150 | MRR@10, Hit Rate@5, Hit Rate@10 | MRR@10 ≥ **0.30** |
+
```
[the remark this answers is not in the document]
Comment (equationalapplications) [reply docs/benchmarks/2026-05-13-financebench-retrieval-benchmark.md:28]: Fixed in 8b42e84: updated benchmark doc — 'Asserted floor' column and Results table threshold both changed from ≥ 0.30 to ≥ 0.55; pass note updated to '1.3× floor' accordingly.

File: packages/integration/__tests__/financebench.test.ts [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    config: { maxResults: 32, tablePrefix: TABLE_PREFIX },
+  });
+  await wiki.setup();
+}, 300_000);
+
+describe('FinanceBench retrieval benchmark', () => {
+  it(
+    'MRR@10 ≥ 0.55 across all 150 FinanceBench questions',
+    async () => {
+      const mrrScores: number[] = [];
```
[the remark this answers is not in the document]
Comment (equationalapplications) [reply packages/integration/__tests__/financebench.test.ts:86]: Fixed in 8b42e84: spec and benchmark docs updated to ≥ 0.55 / maxResults=32 to match the current test. PR description will be updated as part of this /fix-pr follow-up comment.

**Comment by equationalapplications** [comment]
## /fix-pr follow-up

**Commit:** `8b42e84`

### Review resolution

- **Spec threshold mismatch** (`docs/superpowers/specs/2026-05-13-financebench-integration.md`) — **Fixed:** updated Requirement, Goals bullets, code snippet (maxResults 10→32), test name, assertion (≥0.30→≥0.55), and Pass/Fail Threshold table; added note that floor was raised from 0.30 after semantic-32 tuning
- **Benchmark doc threshold mismatch** (`docs/benchmarks/2026-05-13-financebench-retrieval-benchmark.md`) — **Fixed:** "Asserted floor" column and Results table threshold updated from ≥ 0.30 → ≥ 0.55; pass annotation updated to "1.3× floor"
- **PR description / test mismatch** — **Fixed:** spec and benchmark docs now align with the ≥ 0.55 / maxResults=32 CI gate; PR description table shows the current actual value (0.7017) which exceeds the gate

### Verification

- `npm run typecheck` — **pass** (30 files across 4 packages)
- `npm run lint` — **N/A** (no lint script in root package.json)
- `npm run test` — **pass** (302 core + 37 integration + 46 react + 10 expo = 395 tests)

File: packages/integration/__tests__/financebench.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (equationalapplications) [reply packages/integration/__tests__/financebench.test.ts:86]: Fixed: updated PR description to match actual CI gate — MRR@10 ≥ 0.55 with maxResults=32.

File: packages/core/src/WikiMemory.ts [hunk 10]
```diff
@@ -989,7 +989,9 @@ UPDATE ${this.prefix}entries SET source_type = 'librarian_inferred' WHERE source
   async read(entityId: string | string[], query: string, options?: ReadOptions): Promise<MemoryBundle> {
     const config = this.options.config;
     const entityIds = normalizeEntityIds(entityId);
-    const sanitizedTierWeights = sanitizeTierWeights(entityIds, options?.tierWeights);
+    const sanitizedTierWeights = shouldExposeReadMetadata(entityId)
+      ? sanitizeTierWeights(entityIds, options?.tierWeights)
+      : undefined;
     const exposeMetadata = shouldExposeReadMetadata(entityId);
```
[the remark this answers is not in the document]
Comment (equationalapplications) [reply packages/core/src/WikiMemory.ts:995]: Fixed: replaced the inflated Phase 2 section with accurate scope — the actual core change is the dimension-mismatch narrowing to zero-weight entities in WikiMemory.read(), plus two new core test files. The Phase 2 structural retrieval items (Rust indexing, CTE traversal, relational schema, etc.) are not in this PR.

**Comment by equationalapplications** [comment]
## /fix-pr follow-up

**Commit:** no new commit — both issues were PR description only, no code changes required.

### Review resolution

- **Thread / MRR threshold mismatch** (`financebench.test.ts:86`) — **Fixed:** updated PR description MRR@10 threshold from ≥ 0.30 → ≥ 0.55 to match CI gate. Thread resolved.
- **Thread / inflated Phase 2 scope** (`WikiMemory.ts:995`) — **Fixed:** replaced the "Phase 2: WikiMemory structural retrieval" section (which described Rust indexing, CTE traversal, relational schema, UI indicators — none of which are in this PR) with accurate scope: dimension-mismatch narrowing to zero-weight entities + two new core test files. Thread resolved.

### Verification

- `npm run typecheck` — pass (30 files, 0 errors)
- `npm run lint` — N/A (no lint script in repo)
- `npm run test` — pass (395 tests across core/integration/react/expo)
